### ISSUE TITLE
Need a list of supported time string or keyword from date.js

### ISSUE THREAD

**Issue by ttdevelop** [andrewplummer/Sugar#297]
I am trying to apply Date.create(str) in order to parse something like "meeting 5 hours from now" to proper date object, so that requires me to somehow figure out that "5 hours from now" is more valid than "meeting 5". So far I don't know how other people parse it to filter out the time string. If you have some hints, please let me know. Thanks in advance.

**Comment by andrewplummer** [comment]
Sorry this one really fell through the cracks and it's been far too long. For now I should have posted this a long time back http://sugarjs.com/date_formats

I don't know if that's entirely helpful as it's English only at the moment. For other formats I recommend looking at the unit tests as they're a quick reference.

However locales are extremely tailor so it would be quite difficult to make this list exhaustive.

**Comment by andrewplummer** [comment]
So that's, just for a list. However re-reading the use cases here (and looking at issue #475), I'm going to re-examine having a look at "non-strict" parsing, which would essentially do the date parsing without the `^` and `$` tokens to limit the text.

However this would have some big ramifications, both behaviorally and in terms of performance, so it's likely that this will be implemented somehow as an option.

For now it would be nice to have a list to start off the unit tests...

**Comment by andrewplummer** [comment]
Unfortunately after all this time, I've had another look and have only bad news. The solution to the problem as it is currently defined is essentially infeasible. The best way to explain why is to provide an example, although this only scratches the surface. Any solution that would "loosen" the regex requirements essentially boils down to removing the `^` and `$` restrictions to allow date parsing at any point in the string. Here's why that won't work:

```
Date.create('June 15th')
```

The first error (there are many) that you immediately run across is this.  This returns `August 15th` and is matching against a date format that has the regex `/(?:the)?[,. ]*(\d{1,2})(?:st|nd|rd|th)/i)`, which is intended to match ambiguous formats such as "the 15th" or just "15th". When the regex was clamped at both ends there was guaranteed to be no ambiguity in the string. However if we open up the edges, then information meaningful to the date may also be skipped, and a different format matches.

In this case an obvious solution would be to make "the" a non-optional token. However, in addition to disregarding match cases that were previously allowed, there are many other formats where tokens must remain optional. Optional tokens may also exist in front of or behind required tokens, adding to the complexity.

Essentially what happens is that there would be no more guarantee that a format is correctly matching what is intended. A part of the string may match a different format and take on a completely different meaning.

Now, this may be acceptable for you. The formats are ordered to generally follow more specific to more general (however the wildcard here is the optional tokens which throw a wrench into things), so a "best guess" situation may in fact be good enough, even if this sometimes produces errors. If this is the case, then the formats can be forced into a "non-clamping" state using something like this:

```
var en = Date.getLocale('en');
en.getFormats().forEach(function(f) {
  f. reg = RegExp(f.reg.source.replace(/^\^|\$$/g, ''), 'i');
});
// More on this below
en.getFormats = function() {
  return this.compiledFormats;
}
```

This will strip any `^` or `$` from the regexes, fully opening them up. This may work for you or it may entirely cause havoc. I'll leave it to you to decide. However this approach is simply too haphazard a solution to be ever be supported by Sugar.

The second line in the above example is another reason why this is tricky. Essentially, when Sugar successfully matches a date format, it front-loads that format so that successive matches use it first. This can massively increase performance when iterating through many dates with similar formats, as is often the case. The 2nd line above turns off this caching feature, which in theory should produce better results if you do want to try out the above code. However this may drastically reduce performance by turning off the format caching, which is to say nothing of the regex clamping itself (which is also a benefit to performance).

It might be possible to do a deep analysis and produce a "best guess" regex that could at least extract a proper date for ~90% of use cases. This won't be perfect, however, and is something that would require a fairly complex analysis, which means I can basically only do it for English, leaving other locale users out to dry (that is, if it's even possible in other languages to begin with. A "best guess" regex for Japanese, which has no spaces, may not even be in the realm of possibility to begin with). It's also something I would be hesitant about adding back into Sugar, although that would probably depend on the results. Also, even something like this wouldn't be performance-optimal, as you would need to parse out the format once and then run it through `Date.create` to potentially hit all format tokens up as well. But then again I guess if you want to do something that borders on NLP, performance probably isn't your first concern.

TL;DR

I've now considered the methods I've explored for "loosening up" Sugar's date parsing to be dead in the water. I am willing to consider other approaches, however, and one may be a "best guess" regex that can parse out something that is *likely* to be a Sugar parseable token. It won't be perfect but might be good enough.

**Comment by andrewplummer** [comment]
I'm going to close this but added a todo to create better docs for the locales and their formats.
